src/main/webapp/src/app/recommendation/recommendation-form/recommendation-form.component.ts

```diff
@@ -34,13 +37,71 @@ export class RecommendationFormComponent implements OnInit{
     })
 
   }
+
+  ngDoCheck(){
+    if(this.hiddenGems !== this.appComponent.hiddenGems){
+      this.hiddenGems = this.appComponent.hiddenGems;
+    }
+  }
+
   // Submit user's preference form
   // Send hidden gem recommendation as output
-  onSubmit(): void {
-    this.hiddenGemService.findHiddenGemRecommendation(this.preferenceForm.value)
-      .subscribe(hiddenGems => {
-        this.hiddenGemRecommendation.emit(hiddenGems);
-    })
+  async onSubmit() {
+    // filter hidden gems by preference form criteria.
+    let result = this.filterGems(this.hiddenGems, this.preferenceForm);
+
+    // send top 3 recommendation gems to display in list.
+    this.hiddenGemRecommendation.emit(result);
+
+  }
+
+  filterGems(hiddenGems: HiddenGem[], preferenceForm: FormGroup): HiddenGem[] {
+
+    let recommendationGems: HiddenGem[] = hiddenGems.slice(0);
+
+    for (let i = 0; i < recommendationGems.length; i++) {
+      let gem = recommendationGems[i];
+      recommendationGems[i].matchScore = 0;
+      if(preferenceForm.controls["price"].value == recommendationGems[i].priceLevel){
+        recommendationGems[i].matchScore = recommendationGems[i].matchScore + 1;
+
+      }
+      if(preferenceForm.controls["rating"].value >= recommendationGems[i].rating){
+        recommendationGems[i].matchScore = recommendationGems[i].matchScore + 1;
+
+      }
+      if(preferenceForm.controls["type"].value == "any") {
+        recommendationGems[i].matchScore = recommendationGems[i].matchScore + 1;
+
+      } else if (preferenceForm.controls["type"].value == recommendationGems[i].types[0] || preferenceForm.controls["type"].value == recommendationGems[i].types[1]){
```

Comment:
Maybe having the condition as:
recommendationGems[i].types.includes(preferenceForm.controls["type"] 

I am planning to update the backend filtering to take only businesses who are food-related (so having either cafe, bar, restaurant, or meal_takeaway as 1st type), so we should ideally not get a church as recommendation!

Reply:
done